src/Evolution/Systems/GeneralizedHarmonic/Constraints.cpp

```diff
@@ -428,6 +428,75 @@ void two_index_constraint(
       constraint, gamma2, spacetime_normal_one_form, inverse_spacetime_metric,
       spacetime_normal_vector, three_index_constraint);
 }
+
+template <size_t SpatialDim, typename Frame, typename DataType>
+Scalar<DataType> constraint_energy(
+    const tnsr::a<DataType, SpatialDim, Frame>& gauge_constraint,
+    const tnsr::a<DataType, SpatialDim, Frame>& f_constraint,
+    const tnsr::ia<DataType, SpatialDim, Frame>& two_index_constraint,
+    const tnsr::iaa<DataType, SpatialDim, Frame>& three_index_constraint,
+    const tnsr::ijaa<DataType, SpatialDim, Frame>& four_index_constraint,
+    const tnsr::II<DataType, SpatialDim, Frame>& inverse_spatial_metric,
+    double gauge_constraint_multiplier, double two_index_constraint_multiplier,
+    double three_index_constraint_multiplier,
+    double four_index_constraint_multiplier) noexcept {
+  auto energy = make_with_value<Scalar<DataType>>(gauge_constraint, 0.0);
+  constraint_energy<SpatialDim, Frame, DataType>(
+      &energy, gauge_constraint, f_constraint, two_index_constraint,
+      three_index_constraint, four_index_constraint, inverse_spatial_metric,
+      gauge_constraint_multiplier, two_index_constraint_multiplier,
+      three_index_constraint_multiplier, four_index_constraint_multiplier);
+  return energy;
+}
+
+template <size_t SpatialDim, typename Frame, typename DataType>
+void constraint_energy(
+    gsl::not_null<Scalar<DataType>*> energy,
+    const tnsr::a<DataType, SpatialDim, Frame>& gauge_constraint,
+    const tnsr::a<DataType, SpatialDim, Frame>& f_constraint,
+    const tnsr::ia<DataType, SpatialDim, Frame>& two_index_constraint,
+    const tnsr::iaa<DataType, SpatialDim, Frame>& three_index_constraint,
+    const tnsr::ijaa<DataType, SpatialDim, Frame>& four_index_constraint,
+    const tnsr::II<DataType, SpatialDim, Frame>& inverse_spatial_metric,
+    double gauge_constraint_multiplier, double two_index_constraint_multiplier,
+    double three_index_constraint_multiplier,
+    double four_index_constraint_multiplier) noexcept {
+  for (auto& component : *energy) {
```

Comment:
the `for` doesn't seem necessary, and the size isn't checked in the case of a `DataVector` for the contained type. The size checking isn't done in the other constraints, so I'll just make an issue to put it in all of them.